File: bindings/javascript/aal/data/atoms.js
const Aal = require("../gir.js").Aal
const AalWrapper = require("../wrapper.js").AalWrapper
const Atom = require("./atom.js").Atom
const PerAtomProperty = require("./property.js").PerAtomProperty

exports.Atoms = class Atoms extends AalWrapper {
    static sized(size) {
        const obj = new Aal.Atoms.create(size)
        return new Atoms(obj)
    }

    get size() {
        return this._object.getSize()
    }

    set size(val) {
        this._object.setSize(val)
    }

    get(i) {
        if (typeof i === "string") {
            return new PerAtomProperty(this._object.getProp(i))
        } else if (typeof i === "number") {
            return new Atom(this._object.getAtom(i)
        } else {
            throw new Error("NotImplemented")
        }
    }

    set(i, v) {
        if (typeof i === "string") {
            this._object.setProp(i, v._object)
        } else if (typeof i === "number") {
            this._object.setAtom(i, v._object)
        } else {
            throw new Error("NotImplemented")
        }
    }

    del(i) {
        if (typeof i === "string") {
            this._object.delProp(i)
        } else if (typeof i === "number") {
            this._object.delAtom(i)
        } else {
            throw new Error("NotImplemented")
        }
    }

    append(atom) {
        this._object.appendAtom(atom._object)
    }

    keys() {
        return this._object.getPropIds()
    }
}

File: bindings/python/aal/data/atoms.py
from aal.gir import Aal
from aal.wrapper import AalWrapper
from aal.data.atom import Atom
from aal.data.property import PerAtomProperty

class Atoms(AalWrapper):
    SUPPORTED_TYPES = (Aal.Atoms, )
    
    def __init__(self, size: int):
        super().__init__()
        self._object = Aal.Atoms.create(size)

    @property
    def size(self) -> int:
        return self._object.get_size()

    @size.setter
    def size(self, value: int):
        self._object.set_size(value)

    def __getitem__(self, key):
        if isinstance(key, str):
            return PerAtomProperty.from_aal(self._object.get_prop(key))
        elif isinstance(key, int):
            return Atom.from_aal(self._object.get_atom(key))

    def __setitem__(self, key, value):
        if isinstance(key, str):
            self._object.set_prop(key, value._object)
        elif isinstance(key, int):
            self._object.set_atom(key, value._object)

    def __delitem__(self, key):
        if isinstance(key, str):
            self._object.del_prop(key)
        elif isinstance(key, int):
            self._object.del_atom(key)

    def append(self, atom):
        self._object.append_atom(atom._object)

    @property
    def keys(self):
        return self._object.get_prop_ids()
